Template documentation (readme) (#9)
KlausLoeffelmann/NetControlDesigners: src/Resources/Custom TypeEditor Template.md

Custom TypeEditor Template

> Since .NET 3.1 started to support the WinForms Runtime, a new WinForms designer

Reviewer: Hard wrapping makes it harder to review.

> Since .NET 3.1 started to support the WinForms Runtime, a new WinForms designer
> was needed to support .NET applications. The work required a near-complete
> rearchitecting of the designer, as we responded to the differences between .NET
> and the .NET Framework based WinForms designer everyone knows and loves.

Reviewer: ```suggestion
Windows Desktop support was (re-)introduced with .NET Core 3.1, and a new WinForms designer
was needed to support .NET applications. The work required a near-complete
rearchitecting of the designer, as we responded to the differences between .NET
and the .NET Framework based WinForms designer everyone knows and loves.
```

> and the .NET Framework based WinForms designer everyone knows and loves.

Reviewer: I read and re-read this sentence but I'm still having a hard time parsing it, I suggest a re-write/word.
"_The work required a near-complete rearchitecting of the designer, as we responded to the differences between .NET and the .NET Framework based WinForms designer everyone knows and loves._"

Until we added support for .NET Core applications there was only a single
process, devenv.exe, that both the Visual Studio environment and the application
being designed ran within. But .NET Framework and .NET Core can’t both run

> together within devenv.exe, and as a result we had to take the designer out of
> process, thus we called the new designer – WinForms Out of Process Designer (or
> OOP designer for short). We call the existing process – the process that Visual
> Studio runs in – the client process, and the process which shows the actual Form
> at Design time, the server process, which we call *DesignToolsServer*.

Reviewer: ```suggestion
together within devenv.exe, and as a result we had to take the designer out of
process, thus we called the new designer – _WinForms out-of-process designer_. We call the existing process – the process that Visual
Studio runs in – the "client process"; and the process which shows the actual design surface (such as a form or a user control)
at design time - the "server process", which is provided by DesignToolsServer.exe.
```

IIRC, Beth Massi issued a recommendation against the use of "OOP designer".

# Enter the DesignToolsServer

> While simple Control Designer scenarios like Type Converters, Action Lists or
> CodeDom Serializers don’t need any substantial rewrites, Type Editors are a
> different beast altogether.

Reviewer: Unnecessary capitalisation throughout. I'm not going to comment on every instance.
```suggestion
While simple control designer scenarios that use type converters, action lists or
CodeDom serializers may not need any substantial changes, designers that use type editors require significantly more authoring effort.
```

> To illustrate the problem created by introducing this DesignToolsServer as a
> dedicated server process, let’s look at a typical type editor for any image

Reviewer: This kind of reads like "the oil is oily".
```suggestion
To illustrate the problem created by an introduction of a
dedicated server process, let’s look at a typical type editor for any image
```

> To illustrate the problem created by introducing this DesignToolsServer as a
> dedicated server process, let’s look at a typical type editor for any image
> processing control property like the Button’s *BackgroundImage* property: While
> the actual Image that you picked will be rendered on a Button in the Server
> process, the dialog you picked it *with* runs in the context of Visual Studio.

Reviewer: Type (where we refer to an _actual_ type) and properrties should formatted with backticks.

```suggestion
To illustrate the problem created by introducing this DesignToolsServer as a
dedicated server process, let’s look at a type editor for an image property like (e.g., the button’s `BackgroundImage` property).
While the actual image that you pick will be rendered in the server
process, the dialog you picked it *with* runs in the context of Visual Studio (i.e., the client process).
```

That in turn means there is some considerable communication between the two
processes necessary which custom Type Editors for the OOP-Designer need to take
care of. In addition, Type Editors also need to provide a NuGet package which
gets partly loaded into the Visual Studio process and partly into the server
process. And to that end, this NuGet needs to have a special structure (see
below for details). [This blog
post](https://devblogs.microsoft.com/dotnet/state-of-the-windows-forms-designer-for-net-applications/)
describes the concept of the OOP designer and the different processes in greater
detail.

# Projects which the templates create

Setting all these things up manually means coordinating a lot of moving parts,
and there is a huge potential that things go wrong. The individual projects
created by this template help to prevent falling into those traps. The templates
create a series of projects and important Solution Folders, depending on your
needs for both C\# and Visual Basic:

-   **\_Solution Items:** This is a solution folder, which holds this readme,

>     the Directory.Build Target which determines the NuGet package version for

Reviewer: ```suggestion
    the MSBuild props and targets which determines the NuGet package version for
```

    side. On the client, the client-side ViewModel holds the reference to
    server-side view model’s PropertyStore property over a ProxyObject. When the
    user clicks OK in the editor, that code flow is returned to the the
    TypeEditor (running in the context of Visual Studio), which opened the modal
    dialog to begin with. Now, back in the actual type editor class it is where
    the assignment from this ViewModel to the actual Property of the Control
    happens:

var dialogResult = editorService.ShowDialog(_customTypeEditorDialog);

if (dialogResult == DialogResult.OK)

{

// By now, the UI of the Editor has asked its (client-side) ViewModel

// to run the code which updates the property value. It passes the data to

// the server, which in turn updates the server-side ViewModel.

// When it's time to return the value from the client-side ViewModel back to the

// Property Browser (which has called the TypeEditor in the first place), the
client-side

// ViewModel accesses its PropertyStore property, which in turn gets the
required PropertyStore

// proxy object directly from the server-side ViewModel.

value = viewModelClient.PropertyStore;

> }

Reviewer: Broken formatting. Code samples must be wrapped with tripple backticks followed by the language code (`cs` for C# or `vb` for Visual Basic).
